docs/proposals/20230525-openstack-microversion-support.md

```diff
@@ -0,0 +1,95 @@
+
+It is not enough to check if the user has asked for a specific configuration that would require a certain microversion.
+The requirement can come from the OpenStack instance itself.
+As an example, we have seen environments where all volumes are multi-attachable.
+I.e. it is not possible to create a volume that is not multi-attachable, and thus multi-attach support is a must to be able to do anything in this environment, even if the user doesn't explicitly ask for it.
+
+Currently CAPO has a fixed microversion set.
+It is a take it or leave it situation.
+Either this version is supported by the OpenStack environment or you cannot use CAPO.
+
+This proposal is to make CAPO more flexible in what it supports.
+For example, there could be other versions that doesn't really impact CAPO in any way or there could be cross sections between what the OpenStack environment and CAPO supports, that could be useful to the user if certain features are avoided.
+Here is a simple diagram to try to summarize the situation:
+
+```text
+  ┌────────────────────────────────┐
+  │                                │
+  │                                │
+  │     OpenStack support          │
+  │                                │
+  │                                │
+  │                                │
+  │   ┌───────────────────────┐    │
+  │   │                       │    │
+  │   │                       │    │
+  │   │     User usable       │    │
+  │   │                       │    │
+  └───┼───────────────────────┼────┘
+      │                       │
+      │                       │
+      │                       │
+      │                       │
+      │   CAPO support        │
+      │                       │
+      │                       │
+      └───────────────────────┘
+```
+
+### Goals
+
+- Detect user requested features and their version needs.
+
+- Detect server version support.
+
+### Non-goals
+
+- To allow the user to set the versions freely.
+
+### CAPO mocroversions history
+
+Initially no microversions were specified (only versions).
+This is still true for all clients except compute.
+It has been bumped from the original v2 in order to add support for specific features:
+
+- 2.53 is needed for server tags (https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/924)
+- 2.60 is needed for multi-attach volumes (https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1498)
+
+## Proposal
+
+### User stories
+
+#### Story 1
+
+As a user of CAPO, I would like to create a cluster in an older OpenStack environment that doesn't support `currently hard coded microversion` and I am fine with limiting the features available to me based on this older version.
+
+#### Story 2
+
+As a user of CAPO, I would like to create a cluster in an OpenStack environment that requires a higher microversion than `currently hard coded microversion`.
```

Comment:
I don't think this use case exists or is ever likely to exist. I *think* microversions are basically forever.

@stephenfin are you able to confirm/deny?

Reply:
Isn't this exactly why we had https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1498?
We had an OpenStack environment that did not support 2.53 so we had to bump to 2.60 to make CAPO work with it.

Reply:
No. 2.53 was supported, but because multi-attach support was only added in 2.60 (not carefully checking my version numbers here, btw), if you created a volume using microversion 2.53 you continued to get the old behaviour: i.e. no multi-attach. The microversion bump was required to get the new behaviour, but the old one still worked fine.

Reply:
Yes but that is assuming the old behavior was available in the OpenStack environment. I'm saying we saw an environment that straight out rejected 2.53. All volumes *had* to be multi-attach because that was the only thing supported. In other words, it was impossible to use CAPO at all in this environment until the version was bumped to 2.60.

Reply:
Maybe we had different reasons. :shrug:  The message I got on our side was at least that there was some environment where it was impossible to create volumes that wasn't multi-attach, i.e. CAPO was impossible to use without a version bump.

Reply:
I think you're both correct. While there was talk in the past of removing support for older microversions (Ironic were looking into it), I think the conclusion was that there weren't enough benefits to justify it. As such, you should be able to request even the oldest microversion for as long as OpenStack remains a thing.

With that said, if you want to take advantage of newer features then you might have to rely on newer microversions. Typically this would be something service-specific like the ability to indicate a reason when locking a server (introduced in [compute microversion 2.73](https://docs.openstack.org/nova/latest/reference/api-microversion-history.html#id66)) but it could also be something inter-service. A volume supports multi-attach [if the volume type it was created with has the `multiattach` property/extra-spec](https://docs.openstack.org/cinder/latest/admin/volume-multiattach.html). To use such volumes, you must use compute microversion 2.60 or later. If you don't, it'll fail. For example:

```
❯ openstack volume type create --property multiattach='<is> True' mutiattach-demo
❯ openstack volume create --type mutiattach-demo --size 10 my-volume
❯ openstack server create ... --block-device uuid=e3f8fd3f-a380-469f-abde-d0172e54366e,source_type=volume test-server
Multiattach volumes are only supported starting with compute API version 2.60. (HTTP 400) (Request-ID: req-5ffd9c45-ae0e-421e-b89c-0510c85cb3d4)
```

So yes, assuming the default volume type (or perhaps all volume types) has multi-attach enabled then you would be forced to use compute microversion 2.60 to be able to use volumes.

Reply:
Added a note to explain the situation.